File: 프로그래머스/2/42883. 큰 수 만들기/큰 수 만들기.cpp
#include <bits/stdc++.h>
using namespace std;

string solution(string number, int k) {
    list<char> st;

    for (char num : number) {
        while (k > 0 && !st.empty() && st.back() < num) {
            st.pop_back();
            k--;
        }
        st.push_back(num);
    }

    while (k--) st.pop_back();

    string answer = "";
    
    for (auto it = st.begin(); it != st.end(); it++)
        answer += *it;

    return answer;
}

File: 프로그래머스/2/42883. 큰 수 만들기/큰 수 만들기.py
def solution(number, k):
    st = []

    for num in number:
        while k > 0 and st and st[-1] < num:
            st.pop()
            k -= 1
        st.append(num)

    if k > 0:
        st = st[:-k]
    
    return ''.join(st)